File: code/java/NeuralNetwork/src/in/spbhat/NeuralNetwork.java
/*
 * Copyright (c) 2020. Sourabh Bhat ( https://spbhat.in )
 */

package in.spbhat;

import java.util.Random;

import static in.spbhat.util.DoubleArray.apply;
import static in.spbhat.util.DoubleArray.subtract;
import static in.spbhat.util.DoubleMatrix.*;
import static java.lang.Math.exp;

public class NeuralNetwork {
    private static final Random rng = new Random();

    private final int iNodes;
    private final int hNodes;
    private final int oNodes;
    private final double learningRate;
    private double[][] wih;
    private double[][] who;

    public NeuralNetwork(int numInputNodes, int numHiddenNodes, int numOutputNodes, double learningRate) {
        this.iNodes = numInputNodes;
        this.hNodes = numHiddenNodes;
        this.oNodes = numOutputNodes;
        this.learningRate = learningRate;

        // link weight matrices, wih and who
        // weights inside the arrays are wij, where link is from node i to node j in the next layer
        this.wih = createRandom(hNodes, iNodes, 0.0, 1.0 / Math.sqrt(iNodes), rng);
        this.who = createRandom(oNodes, hNodes, 0.0, 1.0 / Math.sqrt(hNodes), rng);
    }

    public void train(double[] inputs, double[] targets) {
        double[] hiddenInputs = multiply(wih, inputs);
        double[] hiddenOutputs = activationFunction(hiddenInputs);
        double[] finalInputs = multiply(who, hiddenOutputs);
        double[] finalOutputs = activationFunction(finalInputs);

        double[] outputErrors = subtract(targets, finalOutputs);
        double[] hiddenErrors = multiply(transpose(who), outputErrors);

        who = add(who, computeWeightsChange(hiddenOutputs, finalOutputs, outputErrors));
        wih = add(wih, computeWeightsChange(inputs, hiddenOutputs, hiddenErrors));
    }

    public double[] query(double[] inputs) {
        double[] hiddenInputs = multiply(wih, inputs);
        double[] hiddenOutputs = apply(hiddenInputs, this::sigmoid);
        double[] finalInputs = multiply(who, hiddenOutputs);

        return apply(finalInputs, this::sigmoid);
    }

    private double[][] computeWeightsChange(double[] inputs, double[] outputs, double[] outputErrors) {
        int numOutputs = outputs.length;

        double[] outputOp = new double[numOutputs];
        for (int i = 0; i < numOutputs; i++) {
            outputOp[i] = learningRate * outputErrors[i] * outputs[i] * (1.0 - outputs[i]);
        }

        return outerProduct(outputOp, inputs);
    }

    private double[][] outerProduct(double[] v1, double[] v2) {
        int rows = v1.length;
        int cols = v2.length;

        double[][] matrix = new double[rows][cols];

        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                matrix[i][j] = v1[i] * v2[j];
            }
        }

        return matrix;
    }

    private double[] activationFunction(double[] x) {
        double[] sigmoid = new double[x.length];
        for (int i = 0; i < x.length; i++) {
            sigmoid[i] = sigmoid(x[i]);
        }
        return sigmoid;
    }

    private double sigmoid(double x) {
        return 1.0 / (1.0 + exp(-x));
    }

    @Override
    public String toString() {
        return "NeuralNetwork{" +
                "iNodes=" + iNodes +
                ", hNodes=" + hNodes +
                ", oNodes=" + oNodes +
                ", learningRate=" + learningRate +
                '}';
    }
}

File: code/python/nn/NeuralNetwork.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Wed Sep 23 07:02:58 2020

@author: Sourabh Bhat ( https://spbhat.in/ )
"""

import numpy as np
import scipy.special

# Neural network class definition
class NeuralNetwork:
    
    # initialize the neural network
    def __init__(self, numInputNodes, numHiddenNodes, numOutputNodes, learningRate):
        # set number of nodes in each input, hidden, output layer
        self.inodes = numInputNodes
        self.hnodes = numHiddenNodes
        self.onodes = numOutputNodes
        
        # learning rate
        self.lr = learningRate
        
        # link weight matrices, wih and who
        # weights inside the arrays are w_i_j, where link is 
        # from node i to node j in the next layer
        # w11 w21
        # w12 w22 etc.
        self.wih = np.random.normal(0.0, 1.0 / np.sqrt(self.inodes),
                                    (self.hnodes, self.inodes))
        self.who = np.random.normal(0.0, 1.0 / np.sqrt(self.hnodes), 
                                    (self.onodes, self.hnodes))
        
        # activation function is the sigmoid function
        self.activation_function = lambda x: scipy.special.expit(x)
        
        pass
    
    # train the neural network
    def train(self, inputs_list, targets_list):
        # convert lists to 2d array
        inputs = np.array(inputs_list, ndmin=2).T
        targets = np.array(targets_list, ndmin=2).T
        
        # calculate signals into hidden layer
        hidden_inputs = np.dot(self.wih, inputs)
        # calculate the signals emerging from hidden layer
        hidden_outputs = self.activation_function(hidden_inputs)
        
        # calculate signals into final output layer
        final_inputs = np.dot(self.who, hidden_outputs)
        # calculate signals emerging from the final output layer
        final_outputs = self.activation_function(final_inputs)
        
        # output layer error is the (target - actual)
        output_errors = targets - final_outputs
        
        # hidden layer error is the output_errors, split by weights 
        # and recombined at hidden nodes
        hidden_errors = np.dot(self.who.T, output_errors)
        
        # update the weights for links between the hidden and output layers
        self.who += self.lr * \
            np.dot(output_errors * final_outputs * (1.0 - final_outputs),
                   hidden_outputs.T)
        
        # update the weights for links between the input and hidden layers
        self.wih += self.lr * \
            np.dot(hidden_errors * hidden_outputs * (1.0 - hidden_outputs),
                   inputs.T)
        
        pass
    
    # query the neural network
    def query(self, inputs_list):
        # convert inputs list to 2d array
        inputs = np.array(inputs_list, ndmin=2).T
        
        # calculate signals into hidden layer
        hidden_inputs = np.dot(self.wih, inputs)
        # calculate the signals emerging from hidden layer
        hidden_outputs = self.activation_function(hidden_inputs)
        
        # calculate signals into final output layer
        final_inputs = np.dot(self.who, hidden_outputs)
        # calculate signals emerging from the final output layer
        final_outputs = self.activation_function(final_inputs)
        
        return final_outputs


if __name__ == "__main__":
    nn = NeuralNetwork(3, 3, 3, 0.3)
    print(nn.query([1, 2, 3]))
    print("test")